### ISSUE TITLE
Handling code versions at SLAC

### ISSUE THREAD

**Issue by SimonKrughoff** [DarkEnergyScienceCollaboration/Twinkles#223]
The most recent run (401) of the twinkles data used a version of the obs_lsstSim code without the saturation interpolation I added at the hack day:
![image](https://cloud.githubusercontent.com/assets/945715/14654625/7c25269e-0633-11e6-9fbf-7bcef544bb8c.png)

I'm not sure how the code is managed at SLAC, but we should probably strive to install packages in the shared stack as the come available and mark them current so it's harder to set up the wrong version.

I'm assuming this is what happened, but please let me know if it's not.

**Comment by jchiang87** [comment]
Can you point to the documentation on how to go about maintaining a "shared" stack?  I've seen different suggestions at community.lsst.org, but nothing definitive.   In any case, I've advertised the existence of new version of the stack [here](https://github.com/DarkEnergyScienceCollaboration/Twinkles/issues/175#issuecomment-211472655) at your request.   The stack location needs to be updated in the workflow code explicitly.

**Comment by SimonKrughoff** [comment]
@jchiang87 I don't know if there is a definitive guide.  I can look around, but thanks for setting up the shared stack as it exists now.

**Comment by tony-johnson** [comment]
@SimonKrughoff I can easily rerun the DMstack workflow with any version of the software that Jim has installed. I interpreted your comments in our phone discussion that you wanted to stick with the original install for now, but I can use the LSST_Stack_2016-02-23 and/or LSST_Stack_2016-04-12 just as easily.

**Comment by jchiang87** [comment]
I would use LSST_Stack_2016-04-12 since it has the fix for the divide by zero and "design matrix" errors.  Based on my tests using supposedly inconsistent astrometry data,  I actually expect the newer stacks to fail at the forced photometry step, so it's a good thing that we have the run with the astrometry fixed using the older stack.

**Comment by SimonKrughoff** [comment]
I don't think we want to use LSST_Stack_2016-04-12 quite yet.  I think I have a fix for the errors @jchiang87 was seeing, but I don't have it pushed to master yet.  Is LSST_Stack_2016-02-23 the one with the hackday saturation fix in it?

**Comment by jchiang87** [comment]
yes. it should be.

**Comment by tony-johnson** [comment]
For Run1.1 we ended up using a version of DMStack installed by @SimonKrughoff. This was great for getting things working, but if we wanted to reproduce this run 6 months from now I am not sure if/how we could reproduce the exact stack build.

My feeling is that we should come up with an automated way of making builds, tracking exactly which code was used. If we want to be able to produce builds which work at multiple sites this needs to be coordinated across site (e.g. #222). We should discuss what are the best tools to handle this.

**Comment by SimonKrughoff** [comment]
I'll note that by using the `lsstsw` mechanism of installing the stack, we should be able to reproduce the environment simply by using the specific build tag used in this run to setup all the packages.  That is not a solution for cross environment tracking.

**Comment by drphilmarshall** [comment]
Code installation gurus: we have an open issue on installing the required version of the DM stack and sims tools needed for the Run 3 PhoSim jobs at SLAC at #319. What do we need to do in order to track well the code used, to enable repeatibility, compatibility with the stack used at NERSC for the DM Level 2 processing, etc, _viz_ the previous discussion we had on this? @tony-johnson ?

**Comment by heather999** [comment]
I'm curious to see some commentary on this and also looking forward to hear how the latest DMstack updates are made available (hopefully published as a release).

In a conda install world, we can make a package list (txt file) to keep track of the DMstack package versions, down to the python modules themselves. That should be reusable between SLAC and NERSC - I think.. those lists can be OS-specific , hence my hesitation.
For a DMStack source build, we can extract versions from the build log or probably via some eups list command.
We'd want to also track phosim and gcc version at least. I'd be interested to see a list of what we want to track actually.

In the short-term, we could just store package lists in Confluence or git..I imagine we intend to use a database ultimately.. what I'd really like is to store a recipe that would allow an easy rebuild of any "release" we use for data runs. I'm not sure how to accomplish that from source using lsstsw or eups distrib install - at least not across SLAC and NERSC such that we obtain an identical build (down to the python modules) to one installed months previously.  I think this level of provenance is still an issue that DM is working on... and of course DMstack isn't our only bit of software to track.
If DMstack/sims started to consistently provide conda installs and if phosim could be persuaded to do the same (or some kind soul provides them with a conda recipe).. could we use conda to handle much of the build management?  Though I do realize then we're subject to the whims of updates to conda itself - which I gather can be problematic.

**Comment by heather999** [comment]
@drphilmarshall getting back to this.. I would like to store the software versions we settle on for Run 3 in a simple Markdown file in the Twinkles repo or in a Confluence page.  It's not a perfect solution, but I am having a hard time following the various issues to keep track of our software decisions.  For fun I might try to write a little script to install those bits if we ever need to replicate this installation.  Does that sound ok for right now? Ultimately we need a more intelligent solution :)

**Comment by drphilmarshall** [comment]
Sounds like a good start to me, Heather! Thanks for all your efforts on
this.
